src/storage/mvcc/txn.rs

```diff
@@ -199,14 +178,40 @@ impl<S: Snapshot> MvccTxn<S> {
         Ok(())
     }
 
+    fn handle_non_pessimistic_lock_conflict(
+        &self,
+        key: Key,
+        for_update_ts: u64,
+        lock: Lock,
+    ) -> Result<()> {
+        // Optimistic lock's for_update_ts is its start_ts.
+        let lock_for_update_ts = if lock.for_update_ts > 0 {
+            lock.for_update_ts
+        } else {
+            lock.ts
+        };
+        if for_update_ts > lock_for_update_ts {
+            Err(Error::KeyIsLocked {
+                key: key.into_raw()?,
+                primary: lock.primary,
+                ts: lock.ts,
```

Comment:
Why not lock.for_update_ts?

Reply:
TiDB needs `start_ts` to resolve the transaction.